### PR TITLE
Merkle trees for commitment to the payset in a block

### PR DESCRIPTION
**Pull request by zeldovich**
This PR describes the specifics of our Merkle tree construction, and describes how that Merkle tree is used to commit to the list of transactions in a block.

### CODE REVIEW

**Comment by tsachiherman** [comment]
I think that our previous language was sufficiently vague to include this feature. But adding this clearly makes it better!

File: dev/crypto.md [hunk 1]
```diff
@@ -91,29 +89,144 @@ for the Byzantine Fault Tolerance protocol, and (2) rerandomize its
...
+```
+def verify(elems, proof, root):
+  if len(elems) == 0:
+    return len(proof) == 0
+  if len(elems) == 1 and len(proof) == 0:
+    return elems[0].pos == 0 && elems[0].hash == root
+
+  i = 0
+  nextelems = []
+  while i < len(elems):
```
Comment (fabrice102) [inline dev/crypto.md:206]: Doesn't this loop infinitely if the condition of the `if` below is not satisfied?
In particular, if `len(elems) = 1`?

File: dev/crypto.md [hunk 2]
```diff
@@ -91,29 +89,144 @@ for the Byzantine Fault Tolerance protocol, and (2) rerandomize its
...
-
-Cryptographic Sortition
------------------------
-
-
 [ledger-spec]: https://github.com/algorand/spec/ledger.md
 [abft-spec]: https://github.com/algorand/spec/abft.md
 
 [sha]: TODO-sha
 [ed25519]: TODO-ed25519
```
Comment (fabrice102) [inline dev/crypto.md:231]: Should https://tools.ietf.org/html/rfc8032 be used as reference for ed25519?

File: dev/crypto.md [hunk 3]
```diff
@@ -91,29 +89,144 @@ for the Byzantine Fault Tolerance protocol, and (2) rerandomize its
...
-
-
-Cryptographic Sortition
------------------------
-
-
 [ledger-spec]: https://github.com/algorand/spec/ledger.md
 [abft-spec]: https://github.com/algorand/spec/abft.md
 
 [sha]: TODO-sha
```
Comment (fabrice102) [inline dev/crypto.md:230]: Should we use https://doi.org/10.6028/NIST.FIPS.180-4?

**Review by fabrice102 — CHANGES_REQUESTED** [review]

File: dev/crypto.md (hunk 1 above)
Comment (zeldovich) [reply dev/crypto.md:206]: Good catch; thanks.  I missed incrementing `i` on each loop iteration of this pseudocode.

File: dev/crypto.md (hunk 2 above)
Comment (zeldovich) [reply dev/crypto.md:231]: Sure, might as well plug it in; thanks.

**Pull request merged**